dhall-lsp-server/src/Dhall/LSP/Backend/Linting.hs

```diff
@@ -22,30 +23,38 @@ data Suggestion = Suggestion {
     }
 
 -- Diagnose nested let blocks.
-diagLetInLet :: Expr Src a -> [Suggestion]
-diagLetInLet (Note _ (Let _ (Note src (Let _ _)))) =
-  [Suggestion (rangeFromDhall src) "Superfluous 'in' before nested let binding"]
-diagLetInLet _ = []
-
--- Given a (noted) let block compute all unused variables in the block.
-unusedBindings :: Eq a => Expr s a -> [Text]
-unusedBindings (Note _ (Let bindings d)) =
+--
+-- Pattern matching on a 'Let' wrapped in a 'Note' prevents us from finding the
+-- same suggestion while searching from multiple 'Let's in the same let-block
+-- – only the outermost 'Let' of a let-block is wrapped in a 'Note'.
+diagLetInLet :: Expr Src a -> Maybe Suggestion
+diagLetInLet (Note _ (Let x mA a b)) = case toMultiLet x mA a b of
+    MultiLet _ (Note src (Let _ _ _ _)) ->
+      Just (Suggestion (rangeFromDhall src) "Superfluous 'in' before nested let binding")
+    _ -> Nothing
+diagLetInLet _ = Nothing
+
+-- Given a let block compute all unused variables in the block.
+unusedBindings :: Eq a => MultiLet s a -> [Text]
+unusedBindings (MultiLet bindings d) =
   let go (Binding var _ _ : []) | not (V var 0 `freeIn` d) = [var]
-      go (Binding var _ _ : (b : bs)) | not (V var 0 `freeIn` Let (b :| bs) d) = [var]
+      go (Binding var _ _ : (b : bs))
+        | let e = fromMultiLet (MultiLet (b :| bs) d)
+        , not (V var 0 `freeIn` e)
+        = [var]
       go _ = []
   in concatMap go (toList $ tails bindings)
-unusedBindings _ = []
 
 -- Diagnose unused let bindings.
 diagUnusedBinding :: Eq a => Expr Src a -> [Suggestion]
-diagUnusedBinding e@(Note src (Let _ _)) = map
+diagUnusedBinding (Note src (Let x mA a e)) = map
   (\var ->
     Suggestion (rangeFromDhall src) ("Unused let binding '" <> var <> "'"))
-  (unusedBindings e)
+  (unusedBindings (toMultiLet x mA a e))
 diagUnusedBinding _ = []
 
 -- | Given an dhall expression suggest all the possible improvements that would
 --   be made by the linter.
 suggest :: Expr Src Import -> [Suggestion]
-suggest expr = concat [ diagLetInLet e ++ diagUnusedBinding e
+suggest expr = concat [ maybeToList (diagLetInLet e) ++ diagUnusedBinding e
```

Comment:
I'm pretty sure that this should produce the same suggestions as before, but I should also check it – I assume there are no automated tests.